packages/remote-feature-flag-controller/src/remote-feature-flag-controller.ts

```diff
@@ -0,0 +1,174 @@
+};
+
+export type RemoteFeatureFlagControllerActions =
+  RemoteFeatureFlagControllerGetStateAction;
+
+export type AllowedActions = never;
+
+export type RemoteFeatureFlagControllerStateChangeEvent =
+  ControllerStateChangeEvent<
+    typeof controllerName,
+    RemoteFeatureFlagControllerState
+  >;
+
+export type RemoteFeatureFlagControllerEvents =
+  RemoteFeatureFlagControllerStateChangeEvent;
+
+export type AllowedEvents = never;
+
+export type RemoteFeatureFlagControllerMessenger =
+  RestrictedControllerMessenger<
+    typeof controllerName,
+    RemoteFeatureFlagControllerActions | AllowedActions,
+    RemoteFeatureFlagControllerEvents | AllowedEvents,
+    AllowedActions['type'],
+    AllowedEvents['type']
+  >;
+
+/**
+ * Returns the default state for the RemoteFeatureFlagController
+ * @returns The default controller state
+ */
+export function getDefaultRemoteFeatureFlagControllerState(): RemoteFeatureFlagControllerState {
+  return {
+    remoteFeatureFlags: {},
+    cacheTimestamp: 0,
+  };
+}
+
+// === CONTROLLER DEFINITION ===
+
+export class RemoteFeatureFlagController extends BaseController<
+  typeof controllerName,
+  RemoteFeatureFlagControllerState,
+  RemoteFeatureFlagControllerMessenger
+> {
+  readonly #fetchInterval: number;
+
+  #disabled: boolean;
+
+  #clientConfigApiService: AbstractClientConfigApiService;
+
+  #inProgressFlagUpdate?: Promise<{ cachedData: FeatureFlags }>;
+
+  constructor({
+    messenger,
+    state,
+    clientConfigApiService,
+    fetchInterval = DEFAULT_CACHE_DURATION,
+    disabled = false,
+  }: {
+    messenger: RemoteFeatureFlagControllerMessenger;
+    state: Partial<RemoteFeatureFlagControllerState>;
+    clientConfigApiService: AbstractClientConfigApiService;
+    fetchInterval?: number;
+    disabled?: boolean;
+  }) {
+    super({
+      name: controllerName,
+      metadata: remoteFeatureFlagControllerMetadata,
+      messenger,
+      state: {
+        ...getDefaultRemoteFeatureFlagControllerState(),
+        ...state,
+      },
+    });
+
+    this.#fetchInterval = fetchInterval;
+    this.#disabled = disabled;
+    this.#clientConfigApiService = clientConfigApiService;
+  }
+
+  private isCacheValid(): boolean {
+    return Date.now() - this.state.cacheTimestamp < this.#fetchInterval;
+  }
+
+  async getRemoteFeatureFlags(): Promise<FeatureFlags> {
+    if (this.#disabled) {
+      return {};
+    }
+
+    if (this.isCacheValid()) {
+      return this.state.remoteFeatureFlags;
+    }
+
+    if (this.#inProgressFlagUpdate) {
+      await this.#inProgressFlagUpdate;
+    }
+
+    try {
+      this.#inProgressFlagUpdate = this.#clientConfigApiService.fetchFlags();
+      const flags = await this.#inProgressFlagUpdate;
+
+      if (Object.keys(flags.cachedData).length > 0) {
+        this.updateCache(flags.cachedData);
+        return flags.cachedData;
+      }
+    } finally {
+      this.#inProgressFlagUpdate = undefined;
+    }
+
+    return this.state.remoteFeatureFlags;
```

Comment:
Thanks for showing me the util, that's exactly what i was looking for to prevent race condition for multiple requests fired :) Addressed in 66ebdbeb6f0108b46c47df2338123cb4d087799b